File: elevenlabs/workflow.js

export class BaseWorkflow {
  constructor(agentWorkflow) {
    this.agentWorkflow = agentWorkflow;
    this.currentWorkflow = null;
    this.currentNode = null;
    this.nextNodeName = null;
    this.currentNodesNameList = [];
  }

  chooseScenario({ scenarioName }) {
    const workflowName = scenarioName;
    if (!this.agentWorkflow.workflows[workflowName]) {
      return `${workflowName} not available, available Scenarios are ${Object.keys(
        this.agentWorkflow.workflows,
      ).join(
        ", ",
      )} Based on the user intent please choose one of the available Scenarios`;
    }

    this.currentWorkflow = this.agentWorkflow.workflows[workflowName];

    this.currentNodesNameList = Object.keys(this.currentWorkflow.nodes);

    this.currentNodeName = this.currentNodesNameList[0];

    this.nextNodeName = this.currentNodesNameList[1] || "";

    this.currentNode = this.currentWorkflow.nodes[this.currentNodeName];

    return this.currentNode.prompt;
  }
}

File: pipecatIntegration/workflow.py
from loguru import logger


class BaseWorkflow:
    def __init__(self, agent_workflow):
        self.agent_workflow = agent_workflow
        self.current_workflow = None
        self.current_node = None
        self.next_node_name = None
        self.current_nodes_name_list = []

    def choose_scenario(self, scenario_name):
        workflows = self.agent_workflow.get("workflows", {})

        if scenario_name not in workflows:
            available = ", ".join(workflows.keys())
            return (
                f"{scenario_name} not available, available Scenarios are {available}. "
                "Based on the user intent please choose one of the available Scenarios"
            )

        self.current_workflow = workflows[scenario_name]
        self.current_nodes_name_list = list(self.current_workflow["nodes"].keys())
        current_node_name = self.current_nodes_name_list[0]
        self.next_node_name = (
            self.current_nodes_name_list[1]
            if len(self.current_nodes_name_list) > 1
            else ""
        )
        self.current_node = self.current_workflow["nodes"][current_node_name]

        logger.info(f"Scenario selected: {scenario_name}")
        return self.current_node["prompt"]

    def get_workflows(self):
        return "\n".join(
            f"{key} - {self.agent_workflow["workflows"][key]["description"]}"
            for key in self.agent_workflow["workflows"].keys()
        )
